Signal/src/util/OWSBackupAPI.swift

```diff
@@ -0,0 +1,245 @@
+                return
+            }
+            Logger.verbose("\(self.logTag) updating record.")
+            let asset = CKAsset(fileURL: fileUrl)
+            record[payloadKey] = asset
+            saveRecordToCloud(record: record,
+                              success: success,
+                              failure: failure)
+        }
+        let myContainer = CKContainer.default()
+        let privateDatabase = myContainer.privateCloudDatabase
+        privateDatabase.add(fetchOperation)
+    }
+
+    @objc
+    public class func saveFileOnceToCloud(recordName: String,
+                                        recordType: String,
+                                        fileUrlBlock: @escaping (Swift.Void) -> URL?,
+                                        success: @escaping (String) -> Swift.Void,
+                                        failure: @escaping (Error) -> Swift.Void) {
+        let recordId = CKRecordID(recordName: recordName)
+        let fetchOperation = CKFetchRecordsOperation(recordIDs: [recordId ])
+        // Don't download the file; we're just using the fetch to check whether or
+        // not this record already exists.
+        fetchOperation.desiredKeys = []
+        fetchOperation.perRecordCompletionBlock = { (record, recordId, error) in
+            if let error = error {
+                if let ckerror = error as? CKError {
+                    if ckerror.code == .unknownItem {
+                        // No record found to update, saving new record.
+
+                        guard let fileUrl = fileUrlBlock() else {
+                            Logger.error("\(self.logTag) error preparing file for upload: \(error).")
+                            return
+                        }
+
+                        saveFileToCloud(fileUrl: fileUrl,
+                                        recordName: recordName,
+                                        recordType: recordType,
+                                        success: success,
+                                        failure: failure)
+                        return
+                    }
+                    Logger.error("\(self.logTag) error fetching record: \(error) \(ckerror.code).")
+                } else {
+                    Logger.error("\(self.logTag) error fetching record: \(error).")
+                }
+                failure(error)
+                return
+            }
+            Logger.info("\(self.logTag) record already exists; skipping save.")
+            success(recordName)
+        }
+        let myContainer = CKContainer.default()
+        let privateDatabase = myContainer.privateCloudDatabase
+        privateDatabase.add(fetchOperation)
+    }
+
+    @objc
+    public class func checkCloudKitAccess(completion: @escaping (Bool) -> Swift.Void) {
+        CKContainer.default().accountStatus(completionHandler: { (accountStatus, error) in
```

Comment:
TODO: observe and handle CKAccountChangedNotification?

From the docs: https://developer.apple.com/documentation/cloudkit/ckcontainer/1399180-accountstatuswithcompletionhandl?language=objc

>  While your app is running, use the CKAccountChangedNotification notification to detect account changes and call this method again to retrieve the status for the new account.

Reply:
Great question, I added it to the design doc.

I suspect we won't need to do this.  The intention is for the UI _and_ our backup jobs to verify cloudkit access when a backup import or export starts (or when backup is enabled).  If a user denied Signal CloudKit/iCloud access mid-restore, the restore would fail (and the user would know why).  If they tried again, they would get an explicit error about CloudKit access.  Similarly, if a background "backup export" job fails to lack of cloudkit access, the user should already be presented with an alert indicating this (although presumably we can improve the copy).